### PR TITLE
Refactoring/gcp/cloudstorage

### PR DESCRIPTION
**Pull request by Remi05**
Migrated GCP Cloud Storage to the new architecture.

### CODE REVIEW

File: ScoutSuite/providers/gcp/facade/cloudstorage.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from google.cloud import storage
+from ScoutSuite.providers.utils import run_concurrently
+
+class CloudStorageFacade:
+    async def get_buckets(self, project_id: str):
+        client = storage.Client(project=project_id)
+        return await run_concurrently(lambda: [bucket for bucket in client.list_buckets()])
```
Comment (misg) [inline ScoutSuite/providers/gcp/facade/cloudstorage.py]:
```suggestion
        return await run_concurrently(lambda: list(client.list_buckets()))
```

File: ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        bucket_dict['versioning_status_enabled'] = raw_bucket.versioning_enabled
+        bucket_dict['logging_enabled'] = self._is_logging_enabled(raw_bucket)
+
+        get_cloudstorage_bucket_acl(raw_bucket, bucket_dict)
+
+        return bucket_dict['id'], bucket_dict
+
+    def _is_logging_enabled(self, raw_bucket):
+        try:
+            return raw_bucket.get_logging() is not None
```
Comment (misg) [inline ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py]: I'm confused about this call. Do you know how it works and if there's an API call behind that?

File: ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        except Exception as e:
+            print_exception('Failed to get bucket logging configuration for %s: %s' % (raw_bucket.name, e))
+            return None # Return False instead?
+
+def get_cloudstorage_bucket_acl(bucket, bucket_dict):
+    try:
+        bucket_acls = bucket.get_iam_policy()
```
Comment (misg) [inline ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py]: Same question as above.

File: ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        get_cloudstorage_bucket_acl(raw_bucket, bucket_dict)
+
+        return bucket_dict['id'], bucket_dict
+
+    def _is_logging_enabled(self, raw_bucket):
+        try:
+            return raw_bucket.get_logging() is not None
+        except Exception as e:
+            print_exception('Failed to get bucket logging configuration for %s: %s' % (raw_bucket.name, e))
+            return None # Return False instead?
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py]: We should address that comment while we're there.  `None` looks good, as long as it is reflected in the UI (with _Unknown_ or something like that).

File: ScoutSuite/providers/gcp/rules/findings/cloudstorage-bucket-no-logging.json [hunk 5]
```diff
@@ -2,9 +2,9 @@
     "dashboard_name": "Buckets",
     "description": "Buckets should have logging enabled",
     "rationale": "By enabling access and storage logs on target Storage buckets, it is possible to capture all events which may affect objects within target buckets (CIS 5.3).",
-    "path": "cloudstorage.buckets.id",
+    "path": "cloudstorage.projects.id.buckets.id",
     "conditions": [ "and",
-        [ "cloudstorage.buckets.id.logging_enabled", "false", "" ]
+        [ "cloudstorage.projects.id.buckets.id.logging_enabled", "false", "" ]
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/rules/findings/cloudstorage-bucket-no-logging.json]: Seems like there might be false negatives here. If get_logging fails, this value will be `None`, and the bucket should be flagged.

```suggestion
        [ "cloudstorage.projects.id.buckets.id.logging_enabled", "false", "" ],
        [ "cloudstorage.projects.id.buckets.id.logging_enabled", "equal", "None" ],

```

File: ScoutSuite/providers/gcp/rules/findings/cloudstorage-bucket-no-logging.json [hunk 6]
```diff
@@ -2,9 +2,9 @@
     "dashboard_name": "Buckets",
     "description": "Buckets should have logging enabled",
     "rationale": "By enabling access and storage logs on target Storage buckets, it is possible to capture all events which may affect objects within target buckets (CIS 5.3).",
-    "path": "cloudstorage.buckets.id",
+    "path": "cloudstorage.projects.id.buckets.id",
     "conditions": [ "and",
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/rules/findings/cloudstorage-bucket-no-logging.json]:
```suggestion
    "conditions": [ "or",
```

File: ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                if member.split(':')[0] not in ['projectEditor', 'projectViewer', 'projectOwner']:
+                    if member not in bucket_dict['acl_configuration']:
+                        bucket_dict['acl_configuration'][member] = [role]
+                    else:
+                        bucket_dict['acl_configuration'][member].append(role)
+        return True
+    except Exception as e:
+        print_exception('Failed to get bucket ACL configuration for %s: %s' % (bucket.name, e))
+        bucket_dict['acls'] = 'Unknown'
+        return False
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py]: The return value isn't used.

```suggestion
```

File: ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        bucket_acls = bucket.get_iam_policy()
+        bucket_dict['acl_configuration'] = {}
+        for role in bucket_acls._bindings:
+            for member in bucket_acls[role]:
+                if member.split(':')[0] not in ['projectEditor', 'projectViewer', 'projectOwner']:
+                    if member not in bucket_dict['acl_configuration']:
+                        bucket_dict['acl_configuration'][member] = [role]
+                    else:
+                        bucket_dict['acl_configuration'][member].append(role)
+        return True
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py]:
```suggestion
```

File: ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py (hunk 2 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py]: There was indeed an API call behind it, I moved it to the facade when initially fetching the buckets.

File: ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py (hunk 3 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/cloudstorage/buckets.py]: Same

**Comment by Remi05** [comment]
I changed a few things regarding fetching of the IAM policy and the logging config so the remaining comments don't really apply anymore. Let me now what you think!

File: ScoutSuite/providers/gcp/facade/cloudstorage.py [hunk 9]
```diff
@@ -1,7 +1,17 @@
...
 
 class CloudStorageFacade:
     async def get_buckets(self, project_id: str):
         client = storage.Client(project=project_id)
-        return await run_concurrently(lambda: list(client.list_buckets()))
+        buckets = await run_concurrently(lambda: list(client.list_buckets()))
+        await asyncio.wait({asyncio.ensure_future(self._add_additional_info(bucket)) for bucket in buckets})
+        return buckets
+
+    async def _add_additional_info(self, bucket):
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/cloudstorage.py]: That is one of the most generic names I've ever seen. Also, I believe the use of a `get_and_set` type method would apply here. See the azure facades, they use that a lot.

File: ScoutSuite/providers/gcp/facade/cloudstorage.py [hunk 10]
```diff
@@ -1,7 +1,17 @@
+import asyncio
 from google.cloud import storage
 from ScoutSuite.providers.utils import run_concurrently
 
 class CloudStorageFacade:
     async def get_buckets(self, project_id: str):
         client = storage.Client(project=project_id)
-        return await run_concurrently(lambda: list(client.list_buckets()))
+        buckets = await run_concurrently(lambda: list(client.list_buckets()))
+        await asyncio.wait({asyncio.ensure_future(self._add_additional_info(bucket)) for bucket in buckets})
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/cloudstorage.py]: I believe the use of a `get_and_set` type method would apply here. See the azure facades, they use that a lot.

Comment (Remi05) [reply ScoutSuite/providers/gcp/facade/cloudstorage.py]: Neat!

**Comment by Remi05** [comment]
Should be ready to merge, but I need to re-test it to make sure. I'm having problems with my GCP account and can't create new resources at the moment. I'll take care of it as soon as the issue with my GCP account gets resolved!

**Pull request merged**